table.cc

```diff
@@ -2453,6 +2453,18 @@ table::run_with_compaction_disabled(std::function<future<> ()> func) {
     });
 }
 
+void
+table::enable_auto_compaction() {
+    _compaction_disabled_by_user = false;
+    // TODO: unmute backlog
+}
```

Comment:
I am thinking about it and dealing with the backlog tracker will be quite hard.
We have to be very careful to:
* not allow it to go negative
* not make it go to zero in the absence of reads and writes for existing compactions
* make it resume from where it stopped if compactions are enabled again.

Having compactions too aggressive in other tables because the backlog is disabled in this table is a small problem compared to any of the above.

This property is an emergency measure, that we don't intend to see used unless there is a problem in the system (but a very important to have one when there is a problem).

I am therefore willing to approve this patch without any of the backlog tracking work.

Just please expand those comments about "mute" and "unmute" a bit to explain to future readers a bit more about what we would have to do if we ever encounter such issue.

Reply:
Ok, I have committed simple implementation that must be workable without backlog muting (as you asked). The current version supports both column family and storage service API. It changes the behavior of CF API, so the patch to the JMX will be required. However, the previous implementation was returning an invalid value anyway.

Having said that, I have the following questions and an example of backlog muting:

> not allow it to go negative

How it can become negative? There is no subtraction anywhere in the backlog tracker impl. 

> not make it go to zero in the absence of reads and writes for existing compactions

Why having backlog == 0 is a bad thing? AFAIU compaction will finish if there are free resources or pause until the backlog will be unmuted(). 

The idea we had about using NCS results in the same backlog == 0 thing.

> make it resume from where it stopped if compactions are enabled again.

Only if we want to track backlog depending on the presented tasks. That is harder than what I have.

So, current PR is ready for the simplest possible impl. Take a look. Also please take a look into this diff that implements backlog muting. I want to understand why this is not an option?

```
diff --git a/sstables/compaction_backlog_manager.hh b/sstables/compaction_backlog_manager.hh
index 01c1d22c1..1d6d2fb7c 100644
--- a/sstables/compaction_backlog_manager.hh
+++ b/sstables/compaction_backlog_manager.hh
@@ -91,6 +91,8 @@ class compaction_backlog_tracker {
     void register_compacting_sstable(sstables::shared_sstable sst, backlog_read_progress_manager& rp);
     void transfer_ongoing_charges(compaction_backlog_tracker& new_bt, bool move_read_charges = true);
     void revert_charges(sstables::shared_sstable sst);
+    void mute() { _mute = true; }
+    void unmute() { _mute = false; }
 private:
     void disable() {
         _disabled = true;
@@ -98,6 +100,7 @@ class compaction_backlog_tracker {
         _ongoing_compactions = {};
     }
     bool _disabled = false;
+    bool _mute = false;
     std::unique_ptr<impl> _impl;
     // We keep track of this so that we can transfer to a new tracker if the compaction strategy is
     // changed in the middle of a compaction.
diff --git a/sstables/compaction_manager.cc b/sstables/compaction_manager.cc
index 7733dbe44..36a4d9ac9 100644
--- a/sstables/compaction_manager.cc
+++ b/sstables/compaction_manager.cc
@@ -793,6 +793,7 @@ void compaction_manager::propagate_replacement(column_family* cf,
 }
 
 double compaction_backlog_tracker::backlog() const {
+    if (_mute) { return 0.0f; }
     return _disabled ? compaction_controller::disable_backlog : _impl->backlog(_ongoing_writes, _ongoing_compactions);
 }
 
diff --git a/table.cc b/table.cc
index c3b36bbf5..69315e32d 100644
--- a/table.cc
+++ b/table.cc
@@ -2455,14 +2455,13 @@ table::run_with_compaction_disabled(std::function<future<> ()> func) {
 
 void
 table::enable_auto_compaction() {
-    // XXX: unmute backlog. turn table backlog back on.
-    //      see table::disable_auto_compaction() notes.
     _compaction_disabled_by_user = false;
+    _compaction_strategy.get_backlog_tracker().mute();
 }
 
 void
 table::disable_auto_compaction() {
-    // XXX: mute backlog. When we disable background compactions
+    // When we disable background compactions
     // for the table, we must also disable current backlog of the
     // table compaction strategy that contributes to the scheduling
     // group resources prioritization.
@@ -2489,6 +2488,7 @@ table::disable_auto_compaction() {
     // - it will break computation of major compaction descriptor
     //   for new submissions
     _compaction_disabled_by_user = true;
+    _compaction_strategy.get_backlog_tracker().unmute();
 }
 
 flat_mutation_reader
```

Reply:
Backlog can go negative because we subtract the contribution of one SSTable when that SSTable is deleted (for instance after a compaction).

If we weren't tracking it in the first place, it can go negative. As a matter of fact, Tomek and I are looking at a bug right now where this happened.


Backlog == 0 is bad because the ongoing compaction may never finish. That's not what we want, if the compaction is already running it should be able to finish. 

Your implementation is actually not that bad. I'd do something different, though, which is consider the backlog muted only when there are no ongoing compactions.

Still, there are potential corner cases that can be lurking in the complexity.
The implementation of the backlog tracker is already quite complex, so I don't think we want to risk making it more complex.

Reply:
> Backlog can go negative because we subtract the contribution of one SSTable when that SSTable is deleted (for instance after a compaction).

> If we weren't tracking it in the first place, it can go negative. 

Thank you! I missed that.

> Backlog == 0 is bad because the ongoing compaction may never finish. That's not what we want, if the compaction is already running it should be able to finish.

Ok. It could be an acceptable tradeoff, but this is understandable.

> I'd do something different, though, which is consider the backlog muted only when there are no ongoing compactions.

I can work on that, but AFAIU you are willing to keep this PR simple?

> The implementation of the backlog tracker is already quite complex, so I don't think we want to risk making it more complex.

Accepted.

Thx for the comments!